Need a short description (OOPS!).

darinadler opened the pull request:
#### f85e0acabccfb2deaea0732156ff5157503f4e4c
<pre>
Need a short description (OOPS!).
Need the bug URL (OOPS!).
Include a Radar link (OOPS!).

Reviewed by NOBODY (OOPS!).

* PerformanceTests/JetStream2/RexBench/OfflineAssembler/LowLevelInterpreter.asm:
* PerformanceTests/JetStream2/RexBench/OfflineAssembler/LowLevelInterpreter.js:
* PerformanceTests/JetStream2/RexBench/OfflineAssembler/expected.js:
* PerformanceTests/RexBench/OfflineAssembler/LowLevelInterpreter.asm:
* PerformanceTests/RexBench/OfflineAssembler/LowLevelInterpreter.js:
* PerformanceTests/RexBench/OfflineAssembler/expected.js:
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::privateCompileMainPass):
(JSC::JIT::privateCompileSlowCases):
* Source/JavaScriptCore/jit/JIT.h:
* Source/JavaScriptCore/jit/JITArithmetic.cpp:
(JSC::JIT::emit_op_less):
(JSC::JIT::emit_op_lesseq):
(JSC::JIT::emit_op_greater):
(JSC::JIT::emit_op_greatereq):
(JSC::JIT::emit_op_jless):
(JSC::JIT::emit_op_jgreatereq):
(JSC::JIT::emitSlow_op_less):
(JSC::JIT::emitSlow_op_lesseq):
(JSC::JIT::emitSlow_op_greater):
(JSC::JIT::emitSlow_op_greatereq):
(JSC::JIT::emit_compare):
(JSC::JIT::emit_compareImpl):
(JSC::JIT::emit_compareSlow):
(JSC::JIT::emit_compareSlowImpl):
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::LLINT_SLOW_PATH_DECL):
* Source/JavaScriptCore/llint/LLIntSlowPaths.h:
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
* Source/JavaScriptCore/runtime/CommonSlowPaths.h:
</pre>

Constellation reviewed (approved):
r=me

cdumez commented on Source/WebCore/html/parser/HTMLToken.h:
```diff
@@ -443,3 +443,11 @@ inline const HTMLToken::Attribute* findAttribute(const HTMLToken::AttributeList&
+
+namespace WTF {
+
+template<> struct VectorTraits<WebCore::HTMLToken::Attribute> : SimpleClassVectorTraits {
```
I don't understand why it is OK to move HTMLToken::Attribute using memcpy. How does this work when the internal vectors have an external buffer?

cdumez replied:
I understand that those 2 vectors have fairly large inline capacity. However, when we add too much data to those vectors, they'd switch to an external VectorBuffer and presumably memcpy wouldn't work anymore?

darinadler replied:
Using `memcpy` for move moves the pointer and the overwrites the original without calling a destructor. It’s true for types like `std::unique_ptr` for that reason (see the `VectorTraits.h` header). There is a separate flag about using `memcpy` for copying that must remain false for the reasons you say above.

weinig replied:
I am not sure you can initialize `HTMLToken::Attribute` with memset since at least the capacity fields in the vectors will then be incorrect (when a Vector has an inline capacity it sets m_capacity = inlineCapacity in the default constructor.

weinig replied:
It does make me wonder if there is a better data structure for the attribute lists I the parser. Maybe just a single UChar buffer with a set pattern:

[number of key / value pairs short][key1 length short][first letter of key1 uchar][second letter of key1 uchar]...[last letter of key1 uchar][value1 length short][first letter of value1 uchar][second letter of value1 uchar]...[last letter of value1 uchar][key2 length short]...

Wrap that in a nice little wrapper that allows appending and forward-only iterating and you got yourself something nice.

darinadler replied:
Yes, thanks @weinig, that’s why the patch doesn’t work.

I’ll look into another solution or maybe just land the traits that improve the performance of moving.

darinadler committed:
Optimize HTML tokenizing by using a custom buffer for attributes instead of a vector of vectors
https://bugs.webkit.org/show_bug.cgi?id=242747

Reviewed by NOBODY (OOPS!).

* Source/WebCore/html/parser/AtomHTMLToken.h:
(WebCore::AtomHTMLToken::initializeAttributes): Update to use functions rather than
data members to get the name and values of attributes.

* Source/WebCore/html/parser/HTMLMetaCharsetParser.cpp:
(WebCore::HTMLMetaCharsetParser::processMeta): Use AtomHTMLToken to convert the
attributs from the token itno a usable form instead of making AtomString directly.
(WebCore::HTMLMetaCharsetParser::encodingFromMetaAttributes): Take a span of
WebCore::Attribute instead of a vector of atom string pairs. A span because
that's the way to pass a vector we just want to read from in new code; more
flexible because we can pass either a vector or any other contiguous memory
like an array. Use the real WebCore::Attribute since it makes it easier to
share code with AtomHTMLToken.
* Source/WebCore/html/parser/HTMLMetaCharsetParser.h: Updated for the above.

* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::StartTagScanner::processAttributes): Pass
the characters from HTMLToken attributes directly to processAttribute,
instead of making an AtomString.
(WebCore::TokenPreloadScanner::StartTagScanner::processAttribute): Make
the AtomString and String out of the attribute name and value only
when needed, rather than always unconditionally doing it.
(WebCore::TokenPreloadScanner::updatePredictedBaseURL): Updated to
use the attribute iterator instead of a findAttribute function.

* Source/WebCore/html/parser/HTMLToken.h: Made AttributeList a custom
class instead of a Vecator<Attribute>. Chose a buffer size of 1000 to
be roughly equivalent to the 10 attributes ecah with a 32-character
name and 64-character value. Also added beginAttributeValue and split
endAttribute into endAttributeWithoutValue and endAttributeWithValue.
Removed now-unneeded m_currentAttribute.

* Source/WebCore/html/parser/HTMLTokenizer.cpp:
(WebCore::HTMLTokenizer::processToken): Added calls to the new
endAttributeWithoutValue and beginAttributeValue functions.
Renamed all the calls of endAttribute to endAttributeWithValue.
Added many missing endAttributeWithValue calls; not entirely
clear why the old code did not assert without these.

* Source/WebCore/page/PageSerializer.cpp:
(WebCore::isCharsetSpecifyingNode): Use Vector<Attribute> instead of
HTMLMetaCharsetParser::AttributeList for compatibility with the changes
to HTMLMetaCharsetParser above.
